File: graph/FloydWarshall/python/FloydWarshall.py
from dataclasses import dataclass
from enum import Enum


I64_MIN = -(1 << 63)
I64_MAX = (1 << 63) - 1


@dataclass(frozen=True)
class Edge:
    source: int
    target: int
    weight: int


class FloydWarshallStatus(Enum):
    OK = "ok"
    NEGATIVE_CYCLE = "negative_cycle"
    OVERFLOW = "overflow"


@dataclass
class FloydWarshallResult:
    status: FloydWarshallStatus
    distances: list[list[int | None]]


def _checked_add(left: int, right: int) -> int | None:
    total = left + right
    if total < I64_MIN or total > I64_MAX:
        return None
    return total


def floyd_warshall(node_count: int, edges: list[Edge]) -> FloydWarshallResult:
    distances: list[list[int | None]] = [[None] * node_count for _ in range(node_count)]

    for node in range(node_count):
        distances[node][node] = 0

    for edge in edges:
        current = distances[edge.source][edge.target]
        if current is None or edge.weight < current:
            distances[edge.source][edge.target] = edge.weight

    for intermediate in range(node_count):
        for source in range(node_count):
            left = distances[source][intermediate]
            if left is None:
                continue

            for target in range(node_count):
                right = distances[intermediate][target]
                if right is None:
                    continue

                candidate = _checked_add(left, right)
                if candidate is None:
                    return FloydWarshallResult(FloydWarshallStatus.OVERFLOW, distances)

                current = distances[source][target]
                if current is None or candidate < current:
                    distances[source][target] = candidate

    for node in range(node_count):
        diagonal = distances[node][node]
        if diagonal is not None and diagonal < 0:
            return FloydWarshallResult(FloydWarshallStatus.NEGATIVE_CYCLE, distances)

    return FloydWarshallResult(FloydWarshallStatus.OK, distances)

File: graph/FloydWarshall/cpp/FloydWarshall.cpp
#include "FloydWarshall.hpp"

#include <limits>

namespace {

bool CheckedAdd(long long left, long long right, long long& out) {
    if (right > 0 && left > std::numeric_limits<long long>::max() - right) {
        return false;
    }

    if (right < 0 && left < std::numeric_limits<long long>::min() - right) {
        return false;
    }

    out = left + right;
    return true;
}

}  // namespace

FloydWarshallResult FloydWarshall(int nodeCount, const std::vector<Edge>& edges) {
    std::vector<std::vector<std::optional<long long>>> distances(
        nodeCount, std::vector<std::optional<long long>>(nodeCount));

    for (int node = 0; node < nodeCount; ++node) {
        distances[node][node] = 0;
    }

    for (const Edge& edge : edges) {
        if (!distances[edge.from][edge.to].has_value() || edge.weight < *distances[edge.from][edge.to]) {
            distances[edge.from][edge.to] = edge.weight;
        }
    }

    for (int intermediate = 0; intermediate < nodeCount; ++intermediate) {
        for (int from = 0; from < nodeCount; ++from) {
            if (!distances[from][intermediate].has_value()) {
                continue;
            }

            for (int to = 0; to < nodeCount; ++to) {
                if (!distances[intermediate][to].has_value()) {
                    continue;
                }

                long long candidate = 0;
                if (!CheckedAdd(*distances[from][intermediate], *distances[intermediate][to], candidate)) {
                    return {FloydWarshallStatus::Overflow, std::move(distances)};
                }

                if (!distances[from][to].has_value() || candidate < *distances[from][to]) {
                    distances[from][to] = candidate;
                }
            }
        }
    }

    for (int node = 0; node < nodeCount; ++node) {
        if (distances[node][node].has_value() && *distances[node][node] < 0) {
            return {FloydWarshallStatus::NegativeCycle, std::move(distances)};
        }
    }

    return {FloydWarshallStatus::Ok, std::move(distances)};
}
